File: Data Structures & Algorithms/balanced-binary-tree/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    return dfs(root) != -1
}

func dfs(node *TreeNode) int {
    if node == nil {
        return 0
    }

    left := dfs(node.Left)
    if left == -1 {
        return -1
    }

    right := dfs(node.Right)
    if right == -1 {
        return -1
    }

    diff := left - right
    if diff > 1 || diff < -1 {
        return -1
    }

    if left > right {
        return left + 1
    }
    return right + 1
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
class Solution {
public:
    int best = INT_MIN;

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return best;
    }

    int dfs(TreeNode* node) {
        if (!node) return 0;
        int left = max(dfs(node->left), 0);
        int right = max(dfs(node->right), 0);
        best = max(best, node->val + left + right);
        return node->val + max(left, right);
    }
};